### ISSUE TITLE
Fine-Tuning 2 Fully Connected Layers (Classification and Regression)

### ISSUE THREAD

**Issue by wzleong** [BVLC/caffe#4927]
I managed to train a model with Regression (Euclidean Loss) and Classification (Softmax). 
I am satisfied with my accuracy for regression but would like to fine-tune my classification. 
Is it possible to just train the classification Fully- Connected layers by setting lr_mult of the Fully-Connected layers responsible for regression as 0? 
***\* Note: 
-  I have a split at the last pooling layer to 2 FC layers (one dedicated for classification and the other for regression)
- I am using Adam update to train the network. 

Thanks 
WZ

**Comment by williford** [comment]
Questions like these should be asked on the [Caffe user mailing list](https://groups.google.com/forum/#!forum/caffe-users) or on another website, such as http://stackoverflow.com/. Please close this issue and move your question to one of these places.

You could also set the weight for the regression loss to be zero.

**Comment by wzleong** [comment]
Hi @williford, 

Thanks for your response I managed to get it working. I am facing another problem and would like to seek for your guidance. 

### Issue summary
I am interested to do some pruning on the trained Caffemodel. I am aware of the tutorial on net surgery. However, I didn't get the results I expected. I am suspecting that the weights and bias didn't get updated into the net. 

### Steps to reproduce

1)  Read parameters from the caffemodel
fc_ori_weights = net.params[fc_layer][0].data
fc_ori_bias = net.params[fc_layer][1].data

2) Apply some modifications on the weights and biases (might even remove some hidden nodes)
- so now fc_weights and fc_bias is updated the shape of the array might be different from the original size
For example: originally fc_weights has size: 128 x 2048
the new fc_weights has size: 89 x 2048
Similarly on bias: 128 --> 89

3) Updating the weights and bias 
net.params[fc_layer][0].flat = fc_weights.flat
net.params[fc_layer][1].flat = fc_bias.flat
net.save('net_surgery.caffemodel')

**Comment by priyapaul** [comment]
@wzleong  how did you do it? Did you manage to find some solution. I have a similar problem

**Comment by wzleong** [comment]
@priyapaul What I did was setting lr_mult = 0 and retrained. Basically it means your learning rate is 0. Hence, no update is applied on the weights.

**Comment by priyapaul** [comment]
Thank you, I have a small question, silly , but I am new to caffe.  If you have a classification and regression task together IN THE SAME NET , how will be the corresponding weights updated? Using accuracy or loss ? Or with respect to both accuracy and loss; SIMULTANEOUSLY while training,and it changes the corresponding weights during back propagation -which is how it is supposed work.For this how should we modify the architecture?  Or you have to run classification and regression separately keeping the weights zero?

**Comment by wzleong** [comment]
Well, what I did was to train either of the network first then extend the network for the other. Yes it updates both weights if you train both together directly. The back propagation or the learning algorithm you favor (SGD or Adam, etc) will work it's way back to update the weights.

**Comment by priyapaul** [comment]
thank you very much!

With Regards

Priya Paul
